spec/duration.html

```diff
@@ -1353,9 +1353,12 @@ <h1>
           1. Set _largestUnit_ to *"hour"*.
         1. Else,
           1. Set _days_ to 0.
+        1. If 𝔽(_days_) is not finite, then
+          1. If _days_ &gt; 0, return ~positive overflow~.
+          1. Else, return ~negative overflow~.
         1. Let _balanceResult_ be ! BalancePossiblyInfiniteTimeDuration(0, 0, 0, 0, 0, 0, _result_.[[Nanoseconds]], _largestUnit_).
         1. If _balanceResult_ is ~positive overflow~ or ~negative overflow~, return _balanceResult_.
-        1. Return ? CreateTimeDurationRecord(_days_, _balanceResult_.[[Hours]], _balanceResult_.[[Minutes]], _balanceResult_.[[Seconds]], _balanceResult_.[[Milliseconds]], _balanceResult_.[[Microseconds]], _balanceResult_.[[Nanoseconds]]).
+        1. Return ! CreateTimeDurationRecord(_days_, _balanceResult_.[[Hours]], _balanceResult_.[[Minutes]], _balanceResult_.[[Seconds]], _balanceResult_.[[Milliseconds]], _balanceResult_.[[Microseconds]], _balanceResult_.[[Nanoseconds]]).
```

Comment:
Oof, I didn't clock that the loop above this could be made infinite with a special calendar/time zone combination.

I _think_ that in built-in time zones and calendars it's only possible for the loop body to execute once, because _days_ would be at most one day longer if _start_ was at, say, 23:59 and _end_ was at 00:00. Is that your understanding as well?

Reply:
The test case for 4edfed17fbfd1af52ffcb729a7267428a382bda8 iterates over the second loop in NanosecondsToDays (step 17). 

The comment you've mentioned from the polyfill implementation of the first loop of NanosecondsToDays (step 13) mentions that transitions >24h are possible: https://github.com/tc39/proposal-temporal/blob/bea08cd3b755d9c4f8df3394b2c47292047ab1cd/polyfill/lib/ecmascript.mjs#L3165-L3172

This doesn't seem to be the case for built-in time zones. Executing this test shows a couple of 24h transitions, but no transitions exceeding 24 hours. (Note: The polyfill implementation shows fewer transitions than the SM/V8 implementations, because transitions before 1847-01-01 are ignored in the polyfill, cf. `BEFORE_FIRST_DST` variable.)
```js
const start = Temporal.Instant.from("1800-01-01T00:00:00Z");
const end = Temporal.Instant.from("2100-01-01T00:00:00Z");
const nanosPerDay = Temporal.Duration.from("P1D").total("nanoseconds");
const oneWeek = Temporal.Duration.from("P7D").round({largestUnit: "hours"});

for (let id of Intl.supportedValuesOf("timeZone")) {
  let tz = new Temporal.TimeZone(id);

  let t = start;
  while (true) {
    let next = tz.getNextTransition(t);
    if (next === null || Temporal.Instant.compare(next, end) >= 0) {
      break;
    }
    t = next;

    // Don't test directly at the transition point.
    let before = tz.getOffsetNanosecondsFor(t.subtract(oneWeek));
    let at = tz.getOffsetNanosecondsFor(t.add(oneWeek));

    let diff = before - at;
    if (diff === nanosPerDay || diff === -nanosPerDay) {
      console.log("24h:", id, next, before, at);
    }
    if (diff > nanosPerDay || diff < -nanosPerDay) {
      console.log(">24h:", id, next, before, at);
    }
  }
}
```

Reply:
I guess the [stock exchange time zone](https://github.com/tc39/proposal-temporal/blob/main/docs/cookbook/stockExchangeTimeZone.mjs) use case would require the ability to have a transition >24h.

Reply:
> I guess the [stock exchange time zone](https://github.com/tc39/proposal-temporal/blob/main/docs/cookbook/stockExchangeTimeZone.mjs) use case would require the ability to have a transition >24h.

This is correct. I think it'd be fine to impose some kind of limit on transitions, but 24 hours seems too short for a limit because it forecloses cases like the stock exchange zone. I could possibly imagine time zones being used for similar cases like school schedules, e.g. grades are assigned to an "instructional day" which can't exist during summer break in the school calendar. 

So if we do impose a limit (which I agree would be a reasonable thing to do), then maybe be conservative and set the limit to something like a year (maybe 500 days to accommodate hypothetical weird lunisolar calendars)?

Reply:
In theory I agree, but now I'm not sure that it really forecloses these cases! I made a local modification which would throw if either of these loops iterated more than once, and the stock exchange time zone continued to work fine. Even when I added more arithmetic:

<details>
<summary>More stock exchange tests</summary>

```js
// Monday lasts 24 hours
const monday = Temporal.ZonedDateTime.from('2022-08-22T09:30[America/New_York]').withTimeZone(tzNYSE);
assert.equal(monday.hoursInDay, 24);

// Friday lasts 72 hours, until Monday starts again
const friday = monday.add({ days: 4 });
assert.equal(friday.hoursInDay, 72);

// Adding 1 day to Friday gets you the next Monday (disambiguates forward)
assert.equal(friday.add({ days: 1 }).toString(), '2022-08-29T09:30:00-04:00[NYSE]');
// Adding 3 days to Friday also gets you the next Monday
assert.equal(friday.add({ days: 3 }).toString(), '2022-08-29T09:30:00-04:00[NYSE]');

const nextMonday = monday.add({ weeks: 1 });

// Subtracting 1 day from Monday gets you the same day (disambiguates forward)
assert.equal(nextMonday.subtract({ days: 1 }).toString(), '2022-08-29T09:30:00-04:00[NYSE]');
// Subtracting 3 days from Monday gets you the previous Friday
assert.equal(nextMonday.subtract({ days: 3 }).toString(), '2022-08-26T09:30:00-04:00[NYSE]');

// Difference between Friday and Monday is 72 hours or 3 days
const fridayUntilMonday = friday.until(nextMonday);
assert.equal(fridayUntilMonday.toString(), 'PT72H');
assert.equal(fridayUntilMonday.total('hours'), 72);
assert.equal(fridayUntilMonday.total('days'), 3);

const mondaySinceFriday = nextMonday.since(friday);
assert.equal(mondaySinceFriday.toString(), 'PT72H');
assert.equal(mondaySinceFriday.total('hours'), 72);
assert.equal(mondaySinceFriday.total('days'), 3);

// One week is still 7 days
const oneWeek = Temporal.Duration.from({ weeks: 1 });
assert.equal(oneWeek.total({ unit: 'days', relativeTo: monday }), 7);
assert.equal(oneWeek.total({ unit: 'days', relativeTo: friday }), 7);
```

</details>

So how about this. I'd be in favour of accepting this PR for now, including the check for infinite days which I'd assume is not normative since as Anba points out, it doesn't affect implementations.

Then, in my normative PR for removing loops and BigInt arithmetic, which I'm planning to open today, I'll put upper limits on these loops that throw if iterated more than once, which is a normative change. If we find a non-contrived use case that throws there, or find that it precludes one of our existing use cases after all, then we can always withdraw that commit from the PR, or put a limit of 365 or 500 instead. Hopefully the public code review gets us many eyes on it :smile:

Reply:
Oh, weird. Not sure why those tests work (and can't check right now) but if they work then I agree with you!

Reply:
I added it to the agenda for the next Temporal meeting - I won't be there though.